src/bulk.ts

```diff
@@ -0,0 +1,176 @@
+import * as prismicT from "@prismicio/types";
+
+/**
+ * The type of a bulk operation.
+ */
+export const BulkOperationType = {
+	CustomTypeInsert: "CUSTOM_TYPE_INSERT",
+	CustomTypeUpdate: "CUSTOM_TYPE_UPDATE",
+	CustomTypeDelete: "CUSTOM_TYPE_DELETE",
+	SliceInsert: "SLICE_INSERT",
+	SliceUpdate: "SLICE_UPDATE",
+	SliceDelete: "SLICE_DELETE",
+} as const;
+export type BulkOperationType =
+	typeof BulkOperationType[keyof typeof BulkOperationType];
+
+/**
+ * An object describing a bulk operation.
+ */
+export type BulkOperation =
+	| {
+			type: typeof BulkOperationType.CustomTypeInsert;
+			id: string;
+			payload: prismicT.CustomTypeModel;
+	  }
+	| {
+			type: typeof BulkOperationType.CustomTypeUpdate;
+			id: string;
+			payload: prismicT.CustomTypeModel;
+	  }
+	| {
+			type: typeof BulkOperationType.CustomTypeDelete;
+			id: string;
+			payload: Pick<prismicT.CustomTypeModel, "id">;
+	  }
+	| {
+			type: typeof BulkOperationType.SliceInsert;
+			id: string;
+			payload: prismicT.SharedSliceModel;
+	  }
+	| {
+			type: typeof BulkOperationType.SliceUpdate;
+			id: string;
+			payload: prismicT.SharedSliceModel;
+	  }
+	| {
+			type: typeof BulkOperationType.SliceDelete;
+			id: string;
+			payload: Pick<prismicT.SharedSliceModel, "id">;
+	  };
+
+export type BulkTransactionModels = {
+	customTypes?: prismicT.CustomTypeModel[];
+	slices?: prismicT.SharedSliceModel[];
+};
+
+const processDiff = <
+	TModel extends prismicT.CustomTypeModel | prismicT.SharedSliceModel,
+>(
+	before: TModel[] = [],
```

Comment:
The default value was added for convenience since arguments like `before.customTypes` could be undefined. I changed `before` and `after` to be required, since it more accurately describes the function's intentions:

```ts
```